src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/MeterListener.cs

```diff
@@ -196,35 +196,35 @@ internal static void NotifyForPublishedInstrument(Instrument instrument)
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal void NotifyMeasurement<T>(Instrument instrument, T measurement, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state)
+        internal void NotifyMeasurement<T>(Instrument instrument, T measurement, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state) where T : struct
         {
-            if (measurement is byte byteMeasurement)
+            if (typeof(T) == typeof(byte))
```

Comment:
@tarekgh Hey I was just noticing this change. Would you indulge me on a couple of questions?

* Why remove the `else if` from the checks?

* I understand the switch from `is` to `typeof(T)` is because some frameworks allocate on `is`. If we added back the old version under a preprocessor directive for >= NET5.0 would it then be allocation free for users on NET5.0 or NET6.0? Right now those users will see at least one boxing from the `object` cast, no?

Reply:
> Why remove the else if from the checks?

It doesn't matter as the JIT optimize the code anyway and will remove all other code when generating type specific code. 

> I understand the switch from is to typeof(T) is because some frameworks allocate on is. If we added back the old version under a preprocessor directive for >= NET5.0 would it then be allocation free for users on NET5.0 or NET6.0? Right now those users will see at least one boxing from the object cast, no?

No, boxing will be optimized away too. no boxing will occur. try it yourself writing some simple code like:

```C#
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        internal static int NotifyMeasurement<T>(T measurement) where T : struct
        {
            if (typeof(T) == typeof(int))
            {
                return (int)(object)measurement;
            }
            if (typeof(T) == typeof(long))
            {
                return (int)(long)(object)measurement;
            }
            return 0;
        }

        [Benchmark]
        public void TestAllocation()
        {
            int t = 0;
            for (int i = 0; i < 10; i++)
            {
                t += NotifyMeasurement<int>(i);
                t += NotifyMeasurement<long>(i);
            }
        }
```

and you will see no allocation happening

```
|         Method |     Mean |     Error |    StdDev | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|----------:|----------:|------:|------:|------:|----------:|
| TestAllocation | 6.096 ns | 0.0659 ns | 0.0584 ns |     - |     - |     - |         - |
```

Reply:
Good stuff thanks @tarekgh! Is this JIT magic documented anywhere? A cheat sheet of things it will optimize away would be great to have.